File: examples/cpp/si_quickstart.cpp
/**
 * @file si_quickstart.cpp
 * @brief iVIT-SDK System Integrator (SI) Quick Start Example
 *
 * Target: System Integrators who need to quickly integrate AI inference
 *         into existing systems.
 *
 * Features demonstrated:
 * - Device discovery and auto-selection
 * - Model loading with load_model()
 * - Structured error handling
 * - Result serialization (JSON)
 *
 * Build:
 *   mkdir build && cd build
 *   cmake .. -DCMAKE_BUILD_TYPE=Release
 *   make si_quickstart
 *
 * Usage:
 *   ./si_quickstart
 *   ./si_quickstart <image_path>
 *   ./si_quickstart <image_path> <model_path>
 */

#include <iostream>
#include <iomanip>
#include <string>
#include <vector>
#include <sstream>
#include <stdexcept>

#include <opencv2/opencv.hpp>

#include "ivit/ivit.hpp"

using namespace ivit;

/**
 * Step 1: Discover available devices
 */
void discover_devices() {
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "Step 1: Device Discovery" << std::endl;
    std::cout << std::string(60, '=') << std::endl;

    // List all available devices
    auto devices = list_devices();
    std::cout << "Found " << devices.size() << " available device(s)" << std::endl;

    for (const auto& device : devices) {
        std::cout << "  - " << device.id << ": " << device.name
                  << " (" << device.backend << ")" << std::endl;
    }

    // Auto-select best device
    auto best = get_best_device();
    std::cout << "\nAuto-selected best device: " << best.id
              << " (" << best.name << ")" << std::endl;

    std::cout << "\nDevice selection options:" << std::endl;
    std::cout << "  get_best_device()                    -> Auto-select best" << std::endl;
    std::cout << "  get_best_device(\"\", \"efficiency\")    -> Best efficiency" << std::endl;
    std::cout << "  DeviceManager::instance().get_device(\"cpu\") -> Specific device" << std::endl;
}

/**
 * Step 2: Load model with error handling
 */
std::shared_ptr<Model> load_model_safe(const std::string& model_path,
                                        const DeviceInfo& device) {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "Step 2: Model Loading" << std::endl;
    std::cout << std::string(60, '=') << std::endl;

    std::cout << "Loading model: " << model_path << std::endl;
    std::cout << "Target device: " << device.id << std::endl;

    LoadConfig config;
    config.device = device.id;

    auto model = load_model(model_path, config);

    std::cout << "Model loaded successfully!" << std::endl;

    // Get model info
    auto input_info = model->input_info();
    if (!input_info.empty()) {
        std::cout << "  Input: " << input_info[0].name
                  << " [" << input_info[0].shape[0];
        for (size_t i = 1; i < input_info[0].shape.size(); ++i) {
            std::cout << ", " << input_info[0].shape[i];
        }
        std::cout << "]" << std::endl;
    }

    return model;
}

/**
 * Step 3: Safe inference result structure
 */
struct InferenceResult {
    bool success;
    std::string error_type;
    std::string message;
    std::string suggestion;
    double inference_time_ms;
    int detection_count;
    std::vector<Detection> detections;

    // Convert to JSON string
    std::string to_json() const {
        std::ostringstream oss;
        oss << "{\n";
        oss << "  \"success\": " << (success ? "true" : "false") << ",\n";

        if (success) {
            oss << "  \"inference_time_ms\": " << std::fixed
                << std::setprecision(2) << inference_time_ms << ",\n";
            oss << "  \"detection_count\": " << detection_count << ",\n";
            oss << "  \"detections\": [\n";

            for (size_t i = 0; i < detections.size(); ++i) {
                const auto& det = detections[i];
                oss << "    {\n";
                oss << "      \"label\": \"" << det.label << "\",\n";
                oss << "      \"confidence\": " << std::fixed
                    << std::setprecision(4) << det.confidence << ",\n";
                oss << "      \"bbox\": [" << det.bbox.x1 << ", " << det.bbox.y1
                    << ", " << det.bbox.x2 << ", " << det.bbox.y2 << "]\n";
                oss << "    }" << (i < detections.size() - 1 ? "," : "") << "\n";
            }

            oss << "  ]\n";
        } else {
            oss << "  \"error_type\": \"" << error_type << "\",\n";
            oss << "  \"message\": \"" << message << "\",\n";
            oss << "  \"suggestion\": \"" << suggestion << "\"\n";
        }

        oss << "}";
        return oss.str();
    }
};

/**
 * Step 3: Safe inference with comprehensive error handling
 */
InferenceResult safe_inference(Model& model, const std::string& image_path) {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "Step 3: Safe Inference" << std::endl;
    std::cout << std::string(60, '=') << std::endl;

    InferenceResult result;

    try {
        std::cout << "Running inference on: " << image_path << std::endl;

        // Load image
        cv::Mat image = cv::imread(image_path);
        if (image.empty()) {
            throw std::runtime_error("Failed to load image: " + image_path);
        }

        // Run inference
        auto results = model.predict(image);

        result.success = true;
        result.inference_time_ms = results.inference_time_ms;
        result.detection_count = static_cast<int>(results.detections.size());
        result.detections = results.detections;

    } catch (const ModelLoadError& e) {
        result.success = false;
        result.error_type = "ModelLoadError";
        result.message = e.what();
        result.suggestion = "Verify model path and format (.onnx, .xml, .engine)";

    } catch (const DeviceNotFoundError& e) {
        result.success = false;
        result.error_type = "DeviceNotFoundError";
        result.message = e.what();
        result.suggestion = "Run list_devices() to check available devices";

    } catch (const InferenceError& e) {
        result.success = false;
        result.error_type = "InferenceError";
        result.message = e.what();
        result.suggestion = "Check input image format and dimensions";

    } catch (const IVITError& e) {
        result.success = false;
        result.error_type = "IVITError";
        result.message = e.what();
        result.suggestion = "See error message for details";

    } catch (const std::exception& e) {
        result.success = false;
        result.error_type = "StdException";
        result.message = e.what();
        result.suggestion = "Unexpected error occurred";
    }

    return result;
}

/**
 * Step 4: Process and display results
 */
void process_results(const InferenceResult& result) {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "Step 4: Results" << std::endl;
    std::cout << std::string(60, '=') << std::endl;

    if (result.success) {
        std::cout << "Inference successful!" << std::endl;
        std::cout << "  Inference time: " << std::fixed << std::setprecision(2)
                  << result.inference_time_ms << " ms" << std::endl;
        std::cout << "  Detections: " << result.detection_count << std::endl;

        for (size_t i = 0; i < result.detections.size(); ++i) {
            const auto& det = result.detections[i];
            std::cout << "    [" << i << "] " << det.label << ": "
                      << std::fixed << std::setprecision(1)
                      << (det.confidence * 100) << "%" << std::endl;
        }

        std::cout << "\nJSON output (for system integration):" << std::endl;
        std::cout << result.to_json() << std::endl;

    } else {
        std::cout << "Inference failed!" << std::endl;
        std::cout << "  Error type: " << result.error_type << std::endl;
        std::cout << "  Message: " << result.message << std::endl;
        std::cout << "  Suggestion: " << result.suggestion << std::endl;
    }
}

void print_best_practices() {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "SI Best Practices:" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "1. Use get_best_device() for auto device selection" << std::endl;
    std::cout << "2. Always wrap inference in try-catch blocks" << std::endl;
    std::cout << "3. Use to_json() for structured output" << std::endl;
    std::cout << "4. Test on multiple hardware environments" << std::endl;
}

int main(int argc, char* argv[]) {
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "iVIT-SDK System Integrator (SI) Quick Start" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "SDK Version: " << ivit::version() << std::endl;

    std::string image_path = argc > 1 ? argv[1] : "test_images/bus.jpg";
    std::string model_path = argc > 2 ? argv[2] : "models/yolov8n.onnx";

    // Step 1: Discover devices
    discover_devices();

    // Get best device
    auto best_device = get_best_device();

    // Step 2: Load model
    std::shared_ptr<Model> model;
    try {
        model = load_model_safe(model_path, best_device);
    } catch (const std::exception& e) {
        std::cerr << "Failed to load model: " << e.what() << std::endl;
        std::cerr << "Please ensure the model file exists." << std::endl;
        return 1;
    }

    // Step 3: Run inference with error handling
    auto result = safe_inference(*model, image_path);

    // Step 4: Process results
    process_results(result);

    // Best practices
    print_best_practices();

    return result.success ? 0 : 1;
}

File: examples/python/advanced/si_quickstart.py
#!/usr/bin/env python3
"""
iVIT-SDK System Integrator (SI) Quick Start Example

Target: System Integrators who need to quickly integrate AI inference
        into existing systems.

Features demonstrated:
- Device discovery and auto-selection
- Model loading (Model Zoo and local files)
- Structured error handling
- Result serialization

Usage:
    python si_quickstart.py
    python si_quickstart.py --image path/to/image.jpg
    python si_quickstart.py --model path/to/model.onnx --image path/to/image.jpg
"""

import argparse
import json
import sys

import numpy as np

import ivit


class NumpyEncoder(json.JSONEncoder):
    """Custom JSON encoder for numpy types."""

    def default(self, obj):
        if isinstance(obj, np.integer):
            return int(obj)
        elif isinstance(obj, np.floating):
            return float(obj)
        elif isinstance(obj, np.ndarray):
            return obj.tolist()
        return super().default(obj)
from ivit.core.exceptions import (
    IVITError,
    ModelLoadError,
    DeviceNotFoundError,
    InferenceError,
    ConfigurationError,
)


def discover_devices():
    """Step 1: Discover available devices."""
    print("=" * 60)
    print("Step 1: Device Discovery")
    print("=" * 60)

    # List all available devices
    devices = ivit.devices()
    print(f"Found {len(devices)} available device(s)")

    # Show device details
    for device in devices:
        print(f"  - {device.id}: {device.name} ({device.backend})")

    # Auto-select best device
    best = ivit.devices.best()
    print(f"\nAuto-selected best device: {best.id} ({best.name})")

    # Alternative: select by type
    print("\nDevice selection options:")
    print(f"  ivit.devices.cpu()  -> CPU device")
    print(f"  ivit.devices.cuda() -> NVIDIA GPU (if available)")
    print(f"  ivit.devices.npu()  -> Intel NPU (if available)")
    print(f"  ivit.devices.best() -> Auto-select best")
    print(f"  ivit.devices.best('efficiency') -> Best efficiency")

    return best


def load_model(model_source: str, device):
    """Step 2: Load model."""
    print("\n" + "=" * 60)
    print("Step 2: Model Loading")
    print("=" * 60)

    if model_source:
        # Load from local file (customer-provided model)
        print(f"Loading model from file: {model_source}")
        model = ivit.load(model_source, device=device)
    else:
        # Load from Model Zoo (quick POC)
        print("Loading model from Model Zoo: yolov8n")
        model = ivit.zoo.load("yolov8n", device=device)

    print(f"Model loaded successfully!")
    print(f"  Device: {model.device}")
    print(f"  Input info: {model.input_info}")

    return model


def safe_inference(model, image_path: str) -> dict:
    """
    Step 3: Safe inference with comprehensive error handling.

    This function demonstrates the recommended error handling pattern
    for production deployments.
    """
    print("\n" + "=" * 60)
    print("Step 3: Safe Inference")
    print("=" * 60)

    try:
        # Execute inference
        print(f"Running inference on: {image_path}")
        results = model(image_path)

        # Get structured results
        output = results.to_dict()

        return {
            "success": True,
            "image_path": image_path,
            "inference_time_ms": results.inference_time_ms,
            "detections": output.get("detections", []),
            "detection_count": len(results),
        }

    except ModelLoadError as e:
        return {
            "success": False,
            "error_type": "ModelLoadError",
            "message": str(e),
            "suggestion": "Please verify model path and format (.onnx, .xml, .engine)"
        }

    except DeviceNotFoundError as e:
        return {
            "success": False,
            "error_type": "DeviceNotFoundError",
            "message": str(e),
            "suggestion": "Run ivit.devices() to check available devices"
        }

    except InferenceError as e:
        return {
            "success": False,
            "error_type": "InferenceError",
            "message": str(e),
            "suggestion": "Check input image format and dimensions"
        }

    except ConfigurationError as e:
        return {
            "success": False,
            "error_type": "ConfigurationError",
            "message": str(e),
            "suggestion": "Check model and device configuration"
        }

    except IVITError as e:
        return {
            "success": False,
            "error_type": "IVITError",
            "message": str(e),
            "suggestion": "See error message for details"
        }

    except FileNotFoundError as e:
        return {
            "success": False,
            "error_type": "FileNotFoundError",
            "message": str(e),
            "suggestion": "Verify the image file path exists"
        }


def process_results(result: dict):
    """Step 4: Process and display results."""
    print("\n" + "=" * 60)
    print("Step 4: Results")
    print("=" * 60)

    if result["success"]:
        print("Inference successful!")
        print(f"  Inference time: {result['inference_time_ms']:.2f} ms")
        print(f"  Detections: {result['detection_count']}")

        for i, det in enumerate(result.get("detections", [])):
            print(f"    [{i}] {det['label']}: {det['confidence']:.2%}")

        # Export as JSON (for system integration)
        print("\nJSON output (for system integration):")
        print(json.dumps(result, indent=2, ensure_ascii=False, cls=NumpyEncoder))

    else:
        print(f"Inference failed!")
        print(f"  Error type: {result['error_type']}")
        print(f"  Message: {result['message']}")
        print(f"  Suggestion: {result['suggestion']}")


def main():
    parser = argparse.ArgumentParser(
        description="iVIT-SDK System Integrator Quick Start"
    )
    parser.add_argument(
        "--model",
        type=str,
        default=None,
        help="Path to model file (.onnx, .xml, .engine). If not provided, uses Model Zoo."
    )
    parser.add_argument(
        "--image",
        type=str,
        default="test_images/bus.jpg",
        help="Path to input image (default: test_images/bus.jpg)"
    )
    args = parser.parse_args()

    print("=" * 60)
    print("iVIT-SDK System Integrator (SI) Quick Start")
    print("=" * 60)
    print(f"SDK Version: {ivit.__version__}")

    # Step 1: Discover devices
    best_device = discover_devices()

    # Step 2: Load model
    try:
        model = load_model(args.model, best_device)
    except Exception as e:
        print(f"\nFailed to load model: {e}")
        print("Please ensure the model file exists or Model Zoo is accessible.")
        sys.exit(1)

    # Step 3: Run inference with error handling
    result = safe_inference(model, args.image)

    # Step 4: Process results
    process_results(result)

    print("\n" + "=" * 60)
    print("SI Best Practices:")
    print("=" * 60)
    print("1. Use ivit.devices.best() for auto device selection")
    print("2. Always wrap inference in try-except blocks")
    print("3. Use results.to_dict() for structured output")
    print("4. Test on multiple hardware environments")


if __name__ == "__main__":
    main()
